File: typescript/src/01-strategy/after/checkout-service.ts
import type { IPaymentStrategy } from './payment-strategy.js';
import type { PaymentResult } from '../../shared/models.js';

// Checkout is decoupled from the gateway. It only knows the contract: .pay(amount).
export class CheckoutService {
  constructor(private readonly payment: IPaymentStrategy) {}

  checkout(amount: number): PaymentResult {
    return this.payment.pay(amount);
  }
}

File: csharp/DesiArchitect.DesignPatterns/01_Strategy/After/CheckoutService.cs
using DesiArchitect.DesignPatterns.Shared;

namespace DesiArchitect.DesignPatterns.Strategy.After;

public class CheckoutService
{
    private readonly IPaymentStrategy _payment;

    public CheckoutService(IPaymentStrategy payment) => _payment = payment;

    public PaymentResult Checkout(decimal amount) => _payment.Pay(amount);
}
